### PR TITLE
Basic Dog Browser

### PR DESCRIPTION
**Pull request by imann24**
**Scripts**
- Added DogBrowser to display all dogs in the game
- Added various UI scripts, mostly for buttons from previous projects
- Added some functions to ArrayUtil
- Improved singleton access to DogDatabase with `GetInstance` auto-initializer upon first access 

**Prefabs**
- Added DogBrowser prefab

**Scenes**
- Added DogBrowserDev scene for testing DogBrowser
- Added DogBrowser to home screen along with "All Dogs" button

**Sprites**
- *James*: Added Various wireframe sprites
- Added 9-Sliced Button Background Sprites

### CODE REVIEW

File: Assets/Scripts/UI/Pickup Pup/DogBrowser.cs [hunk 1]
```diff
@@ -0,0 +1,332 @@
+﻿/*
+ * Author(s): Isaiah Mann
+ * Description: Controls the display of a group of dogs in an organized UI
+ * Usage: Should be attached to the parent element in the UI that holds the child elements
+ */
+
+using UnityEngine;
+
+public class DogBrowser : PPUIElement
```
Comment (gracebarsny) [inline Assets/Scripts/UI/Pickup Pup/DogBrowser.cs]: Do you plan on revising this script at all? It's a bit long. Can come later though.

File: Assets/Scripts/Core/Templates/MonoBehaviourExtended.cs [hunk 2]
```diff
@@ -8,6 +8,9 @@
 
 public abstract class MonoBehaviourExtended : MonoBehaviour, System.IComparable 
 {
+	protected const int NONE_VALUE = 0;
+	protected const int INVALID_VALUE = -1;
+
```
Comment (imann24) [inline Assets/Scripts/Core/Templates/MonoBehaviourExtended.cs]: These aren't meant to cover all edge cases. They're constants to eliminate needing to write -1 and 0 in subclasses. And it reduces the amount of hardcoded vars in the subclasses. I find that it's more expressive, e.g. in a function like:

```
int findIndex(string[] elements, string target)
{
     if(found)
    {
           return index;
     }
     else
     {
         return INVALID_VALUE;
     }
}

File: Assets/Scripts/Data/Core/Database.cs [hunk 3]
```diff
@@ -22,18 +22,44 @@ public static T Instance
 
 	#endregion
 
+	public bool IsInitialized
+	{
+		get
+		{
+			return this.isInitialized;
+		}
```
Comment (imann24) [inline Assets/Scripts/Data/Core/Database.cs]: Felt it was more expressive, but happy to define a convention for use of keyword if we like

File: Assets/Scripts/Data/Core/Database.cs [hunk 4]
```diff
@@ -22,18 +22,44 @@ public static T Instance
...
+	{
+		if(isInitialized)
+		{
+			return false;
+		}
+		else
+		{
+			Initialize();
+			return true;
+		}
```
Comment (imann24) [inline Assets/Scripts/Data/Core/Database.cs]: Generally tend towards using `else` if it makes the statement more expressive

File: Assets/Scripts/Core/Templates/MonoBehaviourExtended.cs (hunk 2 above)
Comment (gracebarsny) [inline Assets/Scripts/Core/Templates/MonoBehaviourExtended.cs]: Great! Thanks for the explanation.

File: Assets/Scripts/UI/Pickup Pup/DogBrowser.cs (hunk 1 above)
Comment (imann24) [inline Assets/Scripts/UI/Pickup Pup/DogBrowser.cs]: This class doesn't strike me as too long to be acceptable. I think 500+ lines is where a class becomes indisputably too long. I could write a child class for handling all the button logic if you think it's worth revising though. Let me know.

Comment (gracebarsny) [inline Assets/Scripts/UI/Pickup Pup/DogBrowser.cs]: Definitely make a separate class for the button logic. We will likely need it for other screens (such as the shop: [https://drive.google.com/file/d/0BxN0RIiq3pWXNnNXaEYwbWl4SFk/]).

In general, if a class is long but not _too_ long, then regions would go a long way.

File: Assets/Scripts/Controller/Pickup Pup/PPGameController.cs [hunk 5]
```diff
@@ -162,6 +164,21 @@ public bool TrySendDogToScout(Dog dog)
...
+	public void SendDogToSlot(Dog dog, DogSlot slot)
+	{
+		slot.Init(dog);
+	}
+
+	public void SendSelectedDogToSlot(DogSlot slot)
+	{
+		SendDogToSlot(this.selectedDog, slot);
+	}
+
```
Comment (imann24) [inline Assets/Scripts/Controller/Pickup Pup/PPGameController.cs]: Will make this one private, but think it's worth keeping for additional functionality

File: Assets/Scripts/UI/Pickup Pup/DogBrowser/DogBrowser.cs [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+		{
+			return dogSlots.Length;
+		}
+	}
+
+
+	[SerializeField]
+	int defaultStartPage;
+
+	DogBrowserButtonController buttonControler;
```
Comment (gracebarsny) [inline Assets/Scripts/UI/Pickup Pup/DogBrowser/DogBrowser.cs]: Spelling error: variable name should be "buttonController"

File: Assets/Scripts/UI/Pickup Pup/DogBrowser/DogBrowserButtonController.cs [hunk 7]
```diff
@@ -0,0 +1,182 @@
+﻿/*
+ * Author(s): Isaiah Mann
+ * Description: Handles the buttons for the DogBrowser
+ * Usage: Should be a one to one relationship between each DogBrowser and its respective ButtonCtroller
+ */
```
Comment (gracebarsny) [inline Assets/Scripts/UI/Pickup Pup/DogBrowser/DogBrowserButtonController.cs]: Spelling: should be "ButtonController"

File: Assets/Scripts/UI/Pickup Pup/DogBrowser/DogBrowserButtonController.cs [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+
+	public void SwitchToPage(int pageIndex, bool onClickPageButton)
+	{
+		if(hasSelectedPage)
+		{
+			// Turn off the last page button 
+			selectedPageButton.Toggle();
+		}
+		selectedPageButton = pageButtons[pageIndex];
+		if (!onClickPageButton)
```
Comment (gracebarsny) [inline Assets/Scripts/UI/Pickup Pup/DogBrowser/DogBrowserButtonController.cs]: Remove space between if and (

**Pull request merged**